core/src/ast/types.rs

```diff
@@ -6,40 +6,57 @@ use lazy_static::lazy_static;
 use std::collections::HashMap;
 use std::iter::FromIterator;
 
-use super::{methods::Method, structs::Struct};
+use super::{
+    methods::Method,
+    structs::{OpaqueStruct, Struct},
+};
 
+/// A type declared inside a Diplomat-annotated module.
 #[derive(Clone, Debug)]
 pub enum CustomType {
+    /// A non-opaque struct whose fields will be visible across the FFI boundary.
     Struct(Struct),
     // TODO(shadaj): Enum
-    Opaque(String, Vec<Method>),
+    /// A struct annotated with [`diplomat::opaque` ]whose fields are not visible.
+    Opaque(OpaqueStruct),
 }
 
 impl CustomType {
+    /// Get the name of the custom type, which is unique within a module.
     pub fn name(&self) -> &String {
         match self {
             CustomType::Struct(strct) => &strct.name,
-            CustomType::Opaque(name, _) => name,
+            CustomType::Opaque(strct) => &strct.name,
         }
     }
 
+    /// Get the methods declared in impls of the custom type.
     pub fn methods(&self) -> &Vec<Method> {
         match self {
             CustomType::Struct(strct) => &strct.methods,
-            CustomType::Opaque(_, methods) => methods,
+            CustomType::Opaque(strct) => &strct.methods,
         }
     }
 }
 
+/// A local type reference, such as the type of a field, parameter, or return value.
+/// Unlike [`CustomType`], which represents a type declaration, [`TypeName`]s can compose
+/// types through references and boxing, and can also capture unresolved paths.
 #[derive(Clone, Debug)]
 pub enum TypeName {
+    /// A built-in Rust scalar primitive.
     Primitive(PrimitiveType),
+    /// An unresolved path to a custom type, which can be resolved after all types
+    /// are collected with [`TypeName::resolve(env)`].
```

Comment:
Does intra-doc links actually ignore things within the parentheses? I don't think it does, probably should just be `TypeName::resolve()`